lib/signalfx.js

```diff
@@ -0,0 +1,334 @@
+
+    if (data === null || data === undefined) {
+      return;
+    }
+
+    var _this = this;
+    for (var metricType in data) {
+      if (data.hasOwnProperty(metricType)) {
+        var dataPoints = data[metricType];
+        if (dataPoints !== null && dataPoints !== undefined) {
+          if (!Array.isArray(dataPoints)) {
+            throw new TypeError('DataPoints not of type list: ' + dataPoints);
+          }
+          dataPoints.forEach(function (dataPoint) {
+            _this.addToQueue(metricType, dataPoint);
+          });
+        }
+      }
+    }
+
+    this.startAsyncSend();
+  };
+
+  SignalFxClient.prototype.addToQueue = function (metricType, datapoint) {
+    throw new Error('Subclasses should implement this!');
+  };
+
+  SignalFxClient.prototype.batchData = function (datapointsList) {
+    throw new Error('Subclasses should implement this!');
+  };
+
+  /**
+   * Send an event to SignalFx
+   *
+   * @param eventType (string): the event type (name of the event time series).
+   * @param dimensions  (dict): a map of event dimensions.
+   * @param properties  (dict): a map of extra properties on that event.
+   */
+  SignalFxClient.prototype.sendEvent = function (eventType, dimensions, properties) {
+    var data = BaseSignalFx.prototype.sendEvent.apply(this, arguments);
+    if (data === null || data === undefined) {
+      return null;
+    }
+
+    var url = [this.apiEndpoint, this.API_ENDPOINT_SUFFIX].join('/');
+    this.post(JSON.stringify(data), url, JSON_HEADER_CONTENT_TYPE);
+  };
+
+  SignalFxClient.prototype.getHeaderContentType = function () {
+    throw new Error('Subclasses should implement this!');
+  };
+
+  SignalFxClient.prototype.startAsyncSend = function () {
+    if (this.asyncRunning) {
+      return;
+    }
+
+    var _this = this;
+    this.asyncRunning = true;
+
+
+    // Send post request in separate thread
+    async.queue(function () {
+      var datapointsList = [];
+      while (_this.queue.length !== 0 && datapointsList.length < _this.batchSize) {
+        datapointsList.push(_this.queue.shift());
+      }
+
+      var dataToSend = _this.batchData(datapointsList);
+      var url = [_this.ingestEndpoint, _this.INGEST_ENDPOINT_SUFFIX].join('/');
+      _this.post(dataToSend, url);
+    }, 2).push();
+  };
+
+  SignalFxClient.prototype.post = function (data, postUrl, contentType) {
+    var _this = this;
+
+    var headers = {};
+    headers[_this.HEADER_USER_AGENT_KEY] = [name, version].join('/');
+    //Adding custom user agents passed by client modules
+    if (_this.userAgents) {
+      headers[_this.HEADER_USER_AGENT_KEY] += ',' + _this.userAgents;
+    }
+    headers[_this.HEADER_API_TOKEN_KEY] = _this.apiToken;
+    headers[_this.HEADER_CONTENT_TYPE] = contentType || _this.getHeaderContentType();
+
+    // An object of options to indicate where to post to
+    var postOptions = {
+      url: postUrl,
+      timeout: this.timeout,
```

Comment:
some parts of this function use `_this` and other times you use `this` (not counting the inner scope within `request` below).